File: artifacts/api-server/src/routes/reports.ts
import { Router, type IRouter } from "express";
import { db, analyticsTable, alertsTable, camerasTable } from "@workspace/db";
import { and, gte, lte, sql } from "drizzle-orm";

const router: IRouter = Router();

router.get("/reports/daily", async (req, res) => {
  const dateStr = req.query.date as string | undefined;
  const targetDate = dateStr ? new Date(dateStr) : new Date();
  targetDate.setHours(0, 0, 0, 0);
  const nextDay = new Date(targetDate.getTime() + 24 * 60 * 60 * 1000);

  const cameras = await db.select().from(camerasTable);
  const cameraMap = Object.fromEntries(cameras.map((c) => [c.id, c.name]));

  const hourExpr = sql`date_trunc('hour', ${analyticsTable.recordedAt})`;

  const rows = await db
    .select({
      timestamp: hourExpr.as("ts"),
      cameraId: analyticsTable.cameraId,
      personCount: sql<number>`sum(${analyticsTable.personCount})::int`,
      compliantCount: sql<number>`sum(${analyticsTable.compliantCount})::int`,
      nonCompliantCount: sql<number>`sum(${analyticsTable.nonCompliantCount})::int`,
      complianceRate: sql<number>`round(avg(${analyticsTable.complianceRate})::numeric, 1)`,
    })
    .from(analyticsTable)
    .where(and(gte(analyticsTable.recordedAt, targetDate), lte(analyticsTable.recordedAt, nextDay)))
    .groupBy(hourExpr, analyticsTable.cameraId)
    .orderBy(hourExpr);

  const alertCount = await db
    .select({ count: sql<number>`count(*)::int` })
    .from(alertsTable)
    .where(and(gte(alertsTable.createdAt, targetDate), lte(alertsTable.createdAt, nextDay)));

  const totalPersonDetections = rows.reduce((s, r) => s + (r.personCount ?? 0), 0);
  const totalNonCompliant = rows.reduce((s, r) => s + (r.nonCompliantCount ?? 0), 0);
  const avgCompliance =
    rows.length > 0
      ? rows.reduce((s, r) => s + (r.complianceRate ?? 0), 0) / rows.length
      : 100;

  const hourlyTotals: Record<string, { personCount: number; hour: string }> = {};
  rows.forEach((r) => {
    const h = String(r.timestamp);
    if (!hourlyTotals[h]) hourlyTotals[h] = { personCount: 0, hour: h };
    hourlyTotals[h].personCount += r.personCount ?? 0;
  });
  const peakEntry = Object.values(hourlyTotals).sort((a, b) => b.personCount - a.personCount)[0];

  const violationsData = await db
    .select({
      hard_hat: sql<number>`sum(${analyticsTable.missingHardHat})::int`,
      safety_vest: sql<number>`sum(${analyticsTable.missingSafetyVest})::int`,
      gloves: sql<number>`sum(${analyticsTable.missingGloves})::int`,
      safety_glasses: sql<number>`sum(${analyticsTable.missingSafetyGlasses})::int`,
      face_mask: sql<number>`sum(${analyticsTable.missingFaceMask})::int`,
      safety_boots: sql<number>`sum(${analyticsTable.missingSafetyBoots})::int`,
    })
    .from(analyticsTable)
    .where(and(gte(analyticsTable.recordedAt, targetDate), lte(analyticsTable.recordedAt, nextDay)));

  const v = violationsData[0] ?? {};
  const topViolations = [
    { ppe: "hard_hat", count: v.hard_hat ?? 0 },
    { ppe: "safety_vest", count: v.safety_vest ?? 0 },
    { ppe: "gloves", count: v.gloves ?? 0 },
    { ppe: "safety_glasses", count: v.safety_glasses ?? 0 },
    { ppe: "face_mask", count: v.face_mask ?? 0 },
    { ppe: "safety_boots", count: v.safety_boots ?? 0 },
  ]
    .filter((x) => x.count > 0)
    .sort((a, b) => b.count - a.count);

  res.json({
    date: targetDate.toISOString().split("T")[0],
    totalPersonDetections,
    uniqueViolations: totalNonCompliant,
    averageComplianceRate: Math.round(avgCompliance * 10) / 10,
    peakHour: peakEntry?.hour ?? null,
    peakPersonCount: peakEntry?.personCount ?? 0,
    alerts: alertCount[0]?.count ?? 0,
    hourlyData: rows,
    topViolations,
  });
});

export default router;

File: artifacts/yolo-service/routes/reports.py
from datetime import datetime, timezone, timedelta
from fastapi import APIRouter, Query
from db import get_cursor

router = APIRouter()


@router.get("/api/reports/daily")
def daily_report(date: str = None):
    if date:
        target_date = datetime.strptime(date, "%Y-%m-%d").replace(tzinfo=timezone.utc)
    else:
        target_date = datetime.now(timezone.utc).replace(hour=0, minute=0, second=0, microsecond=0)
    next_day = target_date + timedelta(days=1)

    with get_cursor() as cur:
        cur.execute("SELECT id, name FROM cameras")
        camera_map = {r["id"]: r["name"] for r in cur.fetchall()}

        cur.execute(
            """SELECT
                date_trunc('hour', recorded_at) AS timestamp,
                camera_id AS "cameraId",
                sum(person_count)::int AS "personCount",
                sum(compliant_count)::int AS "compliantCount",
                sum(non_compliant_count)::int AS "nonCompliantCount",
                round(avg(compliance_rate)::numeric, 1)::float AS "complianceRate"
            FROM analytics
            WHERE recorded_at >= %s AND recorded_at <= %s
            GROUP BY date_trunc('hour', recorded_at), camera_id
            ORDER BY date_trunc('hour', recorded_at)""",
            (target_date, next_day),
        )
        rows = [dict(r) for r in cur.fetchall()]

        for r in rows:
            if r["timestamp"] and hasattr(r["timestamp"], "isoformat"):
                r["timestamp"] = r["timestamp"].isoformat()

        cur.execute(
            "SELECT count(*)::int AS cnt FROM alerts WHERE created_at >= %s AND created_at <= %s",
            (target_date, next_day),
        )
        alert_count = cur.fetchone()["cnt"]

        cur.execute(
            """SELECT
                sum(missing_hard_hat)::int AS hard_hat,
                sum(missing_safety_vest)::int AS safety_vest,
                sum(missing_gloves)::int AS gloves,
                sum(missing_safety_glasses)::int AS safety_glasses,
                sum(missing_face_mask)::int AS face_mask,
                sum(missing_safety_boots)::int AS safety_boots
            FROM analytics
            WHERE recorded_at >= %s AND recorded_at <= %s""",
            (target_date, next_day),
        )
        v = dict(cur.fetchone())

    total_person_detections = sum(r.get("personCount") or 0 for r in rows)
    total_non_compliant = sum(r.get("nonCompliantCount") or 0 for r in rows)
    avg_compliance = (
        sum(r.get("complianceRate") or 0 for r in rows) / len(rows)
        if rows
        else 100
    )

    hourly_totals = {}
    for r in rows:
        h = str(r["timestamp"])
        if h not in hourly_totals:
            hourly_totals[h] = {"personCount": 0, "hour": h}
        hourly_totals[h]["personCount"] += r.get("personCount") or 0
    peak_entry = max(hourly_totals.values(), key=lambda x: x["personCount"]) if hourly_totals else None

    top_violations = [
        {"ppe": k, "count": v.get(k) or 0}
        for k in ["hard_hat", "safety_vest", "gloves", "safety_glasses", "face_mask", "safety_boots"]
        if (v.get(k) or 0) > 0
    ]
    top_violations.sort(key=lambda x: x["count"], reverse=True)

    return {
        "date": target_date.strftime("%Y-%m-%d"),
        "totalPersonDetections": total_person_detections,
        "uniqueViolations": total_non_compliant,
        "averageComplianceRate": round(avg_compliance * 10) / 10,
        "peakHour": peak_entry["hour"] if peak_entry else None,
        "peakPersonCount": peak_entry["personCount"] if peak_entry else 0,
        "alerts": alert_count,
        "hourlyData": rows,
        "topViolations": top_violations,
    }
